### ISSUE TITLE
Allow setting environment variables

### ISSUE THREAD

**Issue by borkdude** [babashka/babashka#849]
Apparently we can do this using `@CFunction, then "like you can do in C"`.

https://www.graalvm.org/sdk/javadoc/org/graalvm/nativeimage/c/function/CFunction.html

https://cornerwings.github.io/2018/07/graal-native-methods/

Working example:

https://gist.github.com/borkdude/335c9911cabf4db6d47cc772b4c69d4d#file-setenv-java-L23-L32

**Comment by borkdude** [comment]
hmm, it seems like you need @CFunction, this allows you call C methods directly from Java code, but bear in mind this will only work in a native-image

Aleksandar Gradinac  3 minutes ago
we have this tutorial https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.tutorial/src/com/oracle/svm/tutorial/CInterfaceTutorial.java for that kind of interop
substratevm/src/com.oracle.svm.tutorial/src/com/oracle/svm/tutorial/CInterfaceTutorial.java
```
/*
* Copyright (c) 2013, 2017, Oracle and/or its affiliates. All rights reserved.
* DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
*
Show more
<https://github.com/oracle/graal|oracle/graal>oracle/graal | Added by GitHub (Legacy)

Aleksandar Gradinac  1 minute ago
basically, you'll need to create a class annotated with CContext , implement CContext.Directives to return <stdlib.h> header for that context, and create a public static native int setenv(CCharPointer name, CCharPointer value, int overwrite); method, annotated with @CFunction (edited) 

Aleksandar Gradinac  < 1 minute ago
you can then use org.graalvm.nativeimage.c.type.CTypeConversion#toCString(java.lang.CharSequence) to convert java strings to C strings, but bear in mind you have to close to holder after you are done with them to not leak memory

**Comment by borkdude** [comment]
For windows, we must follow this workaround:

https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/getenv-wgetenv?view=msvc-160

**Comment by borkdude** [comment]
Note: this doesn't work out that well, since `System/getenv` works with a cached version of the environment properties and from then on they are effectively immutable. You can get around it by using C's `getenv`, but programs that already use `System/getenv` don't see any modifications. This confusion is not worth the effort of adding this to babashka.

Another learning: using GraalVM's C interop has the benefit of yielding fully static binaries for musl, this is not the case with JNI.

**Comment by borkdude** [comment]
Another approach:

```
(set! *warn-on-reflection* true)
(defn set-env [name val]
  (let [env (System/getenv)
        clazz (class env)
        field (.getDeclaredField clazz "m")
        _ (.setAccessible field true)
        the-map (.get field env)
        _ (.put ^java.util.Map the-map name val)]))
(set-env "FOO" "BAR")
```

The output shows that this probably isn't a good idea either:

```
$ clj -M /tmp/setenv.clj
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by user$set_env to field java.util.Collections$UnmodifiableMap.m
WARNING: Please consider reporting this to the maintainers of user$set_env
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```
/cc @fierycod
